Feat(eos_cli_config_gen): Add Monitor Session support

ccsnw opened the pull request:
## Change Summary

- Add option to create monitor sessions

## Related Issue(s)

Related to ##1412

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes

- Proposal for the datamodel below, please review and let me know if that ok 

```
monitor_sessions:
  - name: < session_name_1 >
    sources:
      - name: < interface_name or range >
        direction: < rx | tx | both >
        access_group:
          type: < ip | ipv6 | mac >
          name: < acl_name >
          priority: < priority >
    destinations:
      - < interface(s) | cpu >
    encapsulation_gre_metadata_tx: < true | false >
    header_remove_size: < bytes >
    access_group:
      type: < ip | ipv6 | mac >
      name: < acl_name >
    rate_limit_per_ingress_chip: < "<int> bps" | "<int> kbps" | "<int> mbps" >
    rate_limit_per_egress_chip: < "<int> bps" | "<int> kbps" | "<int> mbps" >
    sample: < int >
    truncate:
      enabled: < true | false >
      size: < bytes >
```

## How to test
added molecule test scenario

## Checklist

### User Checklist

<!-- Add your own checklist using MD syntax and by replacing N/A -->
- N/A

### Repository Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

ccsnw committed:
Feat(eos_cli_config_gen): Add Monitor Session support

ccsnw commented:
Hi guys, would appreciate a review of the datamodel before implementing the rest. 
We should be able to use this in conjunction with a modification in eos_designs to solve: #1412

thanks

ClausHolbechArista commented:
```yaml
monitor_sessions:
  - name: < session_name_1 >
    sources:
      - name: < interface_name or range >
        direction: < rx | tx | both >
        ip_access_group: < acl_name >
        ip_access_group_priority: < priority >
        ipv6_access_group: < acl_name >
        ipv6_access_group_priority: < priority >
        mac_access_group: < acl_name >
        mac_access_group_priority: < priority >
    destinations:
      - < interface(s) | cpu >
    encapsulation_gre_metadata_tx: < true | false >
    header_remove_size: < bytes >
    ip_access_group: < acl_name >
    ip_access_group_priority: < priority >
    ipv6_access_group: < acl_name >
    ipv6_access_group_priority: < priority >
    mac_access_group: < acl_name >
    mac_access_group_priority: < priority >
    rate_limit_per_egress_chip: < "<int> bps" | "<int> kbps" | "<int> mbps" >
    sample: < int >
    truncate:
      enabled: < true | false >
      size: < bytes >
```
This CLI area is highly platform dependent. The supported values and features differ between chips, so I have removed all the specific numbers. 
I have also adjusted a bit to cater for all the knobs, but you don't have to implement all of them. A lof of these are very rarely used, and ex. the source ACLs can be worked around by just appending the commands to the source interface.

ccsnw committed:
Merge branch 'devel' of https://github.com/ccsnw/ansible-avd into monitorsession

ccsnw commented:
> 

Slight adjustment:

```
monitor_sessions:
  - name: < session_name_1 >
    sources:
      - name: < interface_name or range >
        direction: < rx | tx | both >
        ip_access_group: < acl_name >
        ip_access_group_priority: < priority >
        ipv6_access_group: < acl_name >
        ipv6_access_group_priority: < priority >
        mac_access_group: < acl_name >
        mac_access_group_priority: < priority >
    destinations:
      - < interface(s) | cpu >
    encapsulation_gre_metadata_tx: < true | false >
    header_remove_size: < bytes >
    ip_access_group: < acl_name >
    ipv6_access_group: < acl_name >
    mac_access_group: < acl_name >
    rate_limit_per_ingress_chip: < "<int> bps" | "<int> kbps" | "<int> mbps" >
    rate_limit_per_egress_chip: < "<int> bps" | "<int> kbps" | "<int> mbps" >
    sample: < int >
    truncate:
      enabled: < true | false >
      size: < bytes >
```

added: rate_limit_per_ingress_chip
removed: prio values for acls in general config, as far as I can see the only exist per source interface

ccsnw committed:
Feat(eos_cli_config_gen): Add Monitor Session support

ccsnw commented:
@ClausHolbechArista & @tgodaA 

Does it make sense to you to change to sth like:

```
monitor_sessions:
  - name: < session_name_1 >
    sources:
      - name: < interface_name or range >
        direction: < rx | tx | both >
        access_group:
          type: < ip | ipv6 | mac >
          priority: < priority >
    destinations:
      - < interface(s) | cpu >
    encapsulation_gre_metadata_tx: < true | false >
    header_remove_size: < bytes >
    access_group:
      type: < ip | ipv6 | mac >
    rate_limit_per_ingress_chip: < "<int> bps" | "<int> kbps" | "<int> mbps" >
    rate_limit_per_egress_chip: < "<int> bps" | "<int> kbps" | "<int> mbps" >
    sample: < int >
    truncate:
      enabled: < true | false >
      size: < bytes >
```

This might prevent users to input multiple ACLs of different types.. vs. slightly different to acutal CLI

ClausHolbechArista commented:
> @ClausHolbechArista & @tgodaA
> 
> Does it make sense to you to change to sth like:
> 
> ```
> monitor_sessions:
>   - name: < session_name_1 >
>     sources:
>       - name: < interface_name or range >
>         direction: < rx | tx | both >
>         access_group:
>           type: < ip | ipv6 | mac >
>           priority: < priority >
>     destinations:
>       - < interface(s) | cpu >
>     encapsulation_gre_metadata_tx: < true | false >
>     header_remove_size: < bytes >
>     access_group:
>       type: < ip | ipv6 | mac >
>     rate_limit_per_ingress_chip: < "<int> bps" | "<int> kbps" | "<int> mbps" >
>     rate_limit_per_egress_chip: < "<int> bps" | "<int> kbps" | "<int> mbps" >
>     sample: < int >
>     truncate:
>       enabled: < true | false >
>       size: < bytes >
> ```
> 
> This might prevent users to input multiple ACLs of different types.. vs. slightly different to acutal CLI

Good idea. Just remember to also add `name` under `access_group`

ccsnw committed:
Feat(eos_cli_config_gen): Add Monitor Session support

tgodaA commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/monitor-sessions.yml:
```diff
@@ -0,0 +1,40 @@
+      - name: Ethernet12
+        direction: rx
+      - name: Ethernet18
+        access_group: 
```
Remove the trailing space after ```:```.

Please run the following command to ```pre-commit run -a``` to view all lint errors before committing any changes to github.

ccsnw replied:
Thanks Tony, very useful. Adding the check to my routine!

ccsnw committed:
Feat(eos_cli_config_gen): Add Monitor Session support

ccsnw committed:
Feat(eos_cli_config_gen): Add Monitor Session support

ccsnw committed:
Merge branch 'devel' of https://github.com/ccsnw/ansible-avd into monitorsession

ccsnw committed:
Feat(eos_cli_config_gen): Add Monitor Session support

ccsnw commented:
> Good idea. Just remember to also add `name` under `access_group`

@ClausHolbechArista & @tgodaA 

Implemented as discussed and added a first try for the documentation.

My thoughts on it:

- Did not add general session settings explicitly. They are covered by the snippet. Adding them explicitly would most likely be somewhat redundant. Whats your take on it?
- Regarding the table: Decided to but "Access Group" Settings in one column, using "Not defined" in case nothing is defined at all, or priority is not defined. Think replacing these cases with "No Access Group defined" and "No Priority defined" might make more sense. Splitting the three values in a column each seems overkill?

ClausHolbechArista commented:
> * Did not add general session settings explicitly. They are covered by the snippet. Adding them explicitly would most likely be somewhat redundant. Whats your take on it?

You could do a "settings" table like we have under other features, where the columns are "setting" and "value" or similar. Check peer-groups under BGP.

> * Regarding the table: Decided to but "Access Group" Settings in one column, using "Not defined" in case nothing is defined at all, or priority is not defined. Think replacing these cases with "No Access Group defined" and "No Priority defined" might make more sense. Splitting the three values in a column each seems overkill?

I don't think you should build comma-separated fields inside a single table column. IMO it is not making it easier to read.
I suggest that you either create table columns for each value and use `-` as the default value, or write out the cli in the field like "ip access-list blah priority 123" and if priority is not set, just don't print it. Again `-` if no access-list is set.

ccsnw commented:
> > * Did not add general session settings explicitly. They are covered by the snippet. Adding them explicitly would most likely be somewhat redundant. Whats your take on it?
> 
> You could do a "settings" table like we have under other features, where the columns are "setting" and "value" or similar. Check peer-groups under BGP.
> 
> > * Regarding the table: Decided to but "Access Group" Settings in one column, using "Not defined" in case nothing is defined at all, or priority is not defined. Think replacing these cases with "No Access Group defined" and "No Priority defined" might make more sense. Splitting the three values in a column each seems overkill?
> 
> I don't think you should build comma-separated fields inside a single table column. IMO it is not making it easier to read. I suggest that you either create table columns for each value and use `-` as the default value, or write out the cli in the field like "ip access-list blah priority 123" and if priority is not set, just don't print it. Again `-` if no access-list is set.

Thanks Claus.

Let me split the "Access Group" part in columns, it really isn't readable the way it is now.

Regarding the general settings: Still have my doubts about it. We would basically end up with rows like:

```
| Setting | Value |
| Encapsulation Gre Metadata Tx | True |
```
Is that really helpful? If so, should we adjust the model to 

```
monitor_sessions:
  - name: < session_name_1 >
    sources:
      - name: < interface_name or range >
        direction: < rx | tx | both >
        access_group:
          type: < ip | ipv6 | mac >
          name: < acl_name >
          priority: < priority >
    destinations:
      - < interface(s) | cpu >
    session_settings:
      encapsulation_gre_metadata_tx: < true | false >
      header_remove_size: < bytes >
      access_group:
        type: < ip | ipv6 | mac >
        name: < acl_name >
      rate_limit_per_ingress_chip: < "<int> bps" | "<int> kbps" | "<int> mbps" >
      rate_limit_per_egress_chip: < "<int> bps" | "<int> kbps" | "<int> mbps" >
      sample: < int >
      truncate:
        enabled: < true | false >
      size: < bytes >
```

This would allow us to easily check if something under "session_settings" is defined and only render this table if needed? Otherwise we would need to check if anything besides "sources" and "destinations" and "name" exists or always render everything with default settings?

ClausHolbechArista commented:
> Regarding the general settings: Still have my doubts about it. We would basically end up with rows like:
> 
> ```
> | Setting | Value |
> | Encapsulation Gre Metadata Tx | True |
> ```
> 
> Is that really helpful? If so, should we adjust the model to
> 
Creating this table will at least show everything configured. It follows the same style as the rest of the documentation, but I agree it has limited value. Please do the table and we can get this merged :)

> This would allow us to easily check if something under "session_settings" is defined and only render this table if needed? Otherwise we would need to check if anything besides "sources" and "destinations" and "name" exists or always render everything with default settings?

I don't think we should change the model. If you move destination up to the settings table you also avoid repeating the destination in each line in the source table, then the settings table would always be relevant, even with the minimal config.

ccsnw commented:
> If you move destination up to the settings table you also avoid repeating the destination in each line in the source table, then the settings table would always be relevant, even with the minimal config.

This sounds good, haven't thought about it this way. Let me adjust it accordingly and get back to you! 
Thanks!

@ClausHolbechArista changes added. Let me know if it needs further tweaking :-)

ccsnw committed:
Feat(eos_cli_config_gen): Add Monitor Session support

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/monitor-sessions.j2:
```diff
@@ -0,0 +1,55 @@
+{%             endfor %}
+
+##### {{ monitor_session.name }} Destinations and Session Settings
+| Settings | Values |
```
missing one blank line between heading and table. Same above

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/monitor-sessions.j2:
```diff
@@ -0,0 +1,55 @@
+## Monitor Sessions
+
+### Monitor Sessions Summary
+
```
move this blank line down to just above the `#### {{ monitor_session.name }}`. This way it will produce a blank line for every loop, but none if the list is empty.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/monitor-sessions.j2:
```diff
@@ -0,0 +1,55 @@
+| Truncate Enabled | {{ monitor_session.truncate.enabled }} |
+{%             endif %}
+{%         endif %}
+
```
Move this blank line down just above the `### Monitor Sessions Configuration`

ccsnw committed:
Merge branch 'devel' of https://github.com/ccsnw/ansible-avd into monitorsession

ccsnw committed:
Feat(eos_cli_config_gen): Add Monitor Session support

ClausHolbechArista reviewed (approved):
LGTM

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/monitor-sessions.j2:
```diff
@@ -0,0 +1,57 @@
+{%             if monitor_session.sample is arista.avd.defined %}
+| Sample | {{ monitor_session.sample }} |
+{%             endif %}
+{%             if monitor_session.truncate.enabled is arista.avd.defined(true) and monitor_session.truncate.size is arista.avd.defined %}
```
Looks good. I have a suggestion for lines 43-48:

```
{%             if monitor_session.truncate.enabled is arista.avd.defined(true) %}
| Truncate Enabled | {{ monitor_session.truncate.enabled }} |
{%                 if monitor_session.truncate.size is arista.avd.defined %}
| Truncate Size | {{ monitor_session.truncate.size }} |
{%                 endif %}
{%             endif %}
```

ccsnw replied:
thanks!! i'm still thinking too complicated sometimes

tgodaA replied:
Our brain always wants something complex! :)

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/README.md:
```diff
@@ -2075,6 +2076,33 @@ snmp_server:
+          name: < acl_name >
+          priority: < priority >
+    destinations:
+      - < interface(s) | cpu >
```
I have no idea about this. So just wanted to know. Does this knob support any type of interfaces (vlan, lo) or only Ethernet ?

ccsnw replied:
commonly ethernet interfaces, portchannels or CPU

apparently some platforms also suppor to mirror to a gre tunnel:

https://www.arista.com/en/um-eos/eos-data-transfer#xx1135616

tgodaA commented on ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/monitor-sessions.yml:
```diff
@@ -0,0 +1,38 @@
+    sources:
+      - name: Ethernet12
+        direction: rx
+      - name: Ethernet18
```
Not a suggestion but provide an example for interface_range if possible. So that we cover that test case as well.

You can edit this one or add a new test case.

ccsnw committed:
Feat(eos_cli_config_gen): Add Monitor Session support

ccsnw committed:
Feat(eos_cli_config_gen): Add Monitor Session support

tgodaA reviewed (approved):
LGTM!
